### ISSUE TITLE
add TaskGroup here or in another library

### ISSUE THREAD

**Issue by dhalbert** [adafruit/Adafruit_CircuitPython_asyncio#7]
asyncio could really use a `TaskGroup` implementation, as provided by [Trio](https://trio.readthedocs.io/en/stable/) (aka nurseries), [Curio](https://curio.readthedocs.io/en/stable/), [aiotools](https://aiotools.readthedocs.io/en/latest/aiotools.taskgroup.html), and [EdgeDB](https://github.com/edgedb/edgedb/blob/master/edb/common/taskgroup.py).

`TaskGroup` was originally slated to be added Python as long ago as 3.8, but was held up because exception group handling was not available. However, exception groups are now being added: [PEP 654](https://www.python.org/dev/peps/pep-0654/), [PR](https://github.com/python/cpython/pull/29581)). Once that PEP is implemented, `TaskGroup` will probably follow. But the timeline for PEP 654 to go into the MicroPython core is unclear. So any TaskGroup we add will not be quite compatible, and we probably want to add it in an ancillary library (`adafruit_asyncio` or `asynciox` or `xasyncio` or ??).

**Comment by dhalbert** [comment]
TaskGroups going in to CPython 3.11:
https://bugs.python.org/issue46752
https://github.com/python/cpython/pull/31270

Quattro version (not using multi-except):
https://github.com/Tinche/quattro

**Comment by dorinclisu** [comment]
I agree it would be a good idea to support structured concurrency, as going head-on with the older asyncio primitives can get pretty messy.
